vendors/nuvoton/boards/numaker_iot_m487_wifi/aws_demos/application_code/nuvoton_code/entropy_hardware_poll.c

```diff
@@ -0,0 +1,221 @@
+/*
+ *  Hardware entropy collector for M487 RNG
+ *
+ */
+
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include "FreeRTOS.h"
+#include "task.h"
+#include "NuMicro.h"
+
+/*
+ * Get Random number generator.
+ */
+ #define PRNG_KEY_SIZE  (0x20UL)
+ 
+static volatile int  g_PRNG_done;
+volatile int  g_AES_done;
+
+
+#define SNUM        32       /* recorded number of lastest samples */
+
+
+static uint32_t   adc_val[SNUM];
+static uint32_t   val_sum;
+static int        oldest;
+
+#ifdef __ICCARM__
+#define __inline   inline
+#endif
+
+static __inline uint32_t  get_adc_bg_val()
+{
+    EADC->SWTRG = (1 << 16);      //Trigger Band-gap
+    while ((EADC->STATUS1 & 1) != 1);
+    return (EADC->DAT[16] & 0xFFF);
+}
+
+int  adc_trng_gen_bit()
+{
+    uint32_t   new_val, average;
+    int        ret_val;
+    int ref_bg_ub = ((1.2/3.3)*4095) + 50;
+    int ref_bg_lb = ((1.2/3.3)*4095) - 50;
+    
+    do{
+      new_val = get_adc_bg_val();
+    } while( (new_val > ref_bg_ub) || (new_val < ref_bg_lb) );
+
+    average = (val_sum / SNUM);   /* sum divided by 32 */
+
+    if (average >= new_val)
+        ret_val = 1;
+    else
+        ret_val = 0;
+
+    //printf("%d - sum = 0x%x, avg = 0x%x, new = 0x%x\n", oldest, val_sum, average, new_val);
+
+    /* kick-off the oldest one and insert the new one */
+    val_sum -= adc_val[oldest];
```

Comment:
This global is accessed without locking. I suspect that if there are multiple tasks with a P#11 RNG context, these store/load instructions could race.

Reply:
@dcgaws , yes, I will implement mutex lock for multiple task case.

Reply:
@dcgaws , finished the protection by mutex in commit  36035e015  [the-repo](https://github.com/cyliangtw/amazon-freertos/tree/feature/m487_cmake_0531)